### PR TITLE
Feat (Curve): Detect prices of LP and Gauge tokens

### PR DESCRIPTION
**Pull request by OjusWiZard**
## Checklist

- [x] Detect price of LP and Gauge tokens of Curve on all supported chains

### CODE REVIEW

File: rotkehlchen/chain/evm/node_inquirer.py [hunk 1]
```diff
@@ -173,7 +173,59 @@ def _query_web3_get_logs(
...
+                'database': self.database,
+                'new_data': new_data,
+            }
+            if chain_id is not None:
+                save_method_kwargs['chain_id'] = chain_id
+            save_method(**save_method_kwargs)
+        return True
+
+
+class EvmNodeInquirer(ABC, UpdatableCacheDataMixin):
```
Comment (LefterisJP) [inline rotkehlchen/chain/evm/node_inquirer.py:228]: if you do this then:

1. Is it used **anywhere** else? If not then why not just make it part of evm node inquirer itself?
2. Everywhere else that UpdatableCacheDataMixin is used should be removed. Atm every chain Inquirer inherits from it. This should be undone. Grep for uses.

File: rotkehlchen/rotkehlchen.py [hunk 2]
```diff
@@ -426,6 +426,10 @@ def unlock_user(
         Inquirer().inject_evm_managers([
             (ChainID.ETHEREUM, ethereum_manager),
             (ChainID.OPTIMISM, optimism_manager),
+            (ChainID.POLYGON_POS, polygon_pos_manager),
+            (ChainID.ARBITRUM_ONE, arbitrum_one_manager),
+            (ChainID.BASE, base_manager),
+            (ChainID.GNOSIS, gnosis_manager),
```
Comment (LefterisJP) [inline rotkehlchen/rotkehlchen.py:432]: What about scroll? Should it be all here? If yes then we need this to be an iteration of the `EVM_CHAINS_WITH_TRANSACTIONS`

File: rotkehlchen/tests/unit/test_inquirer.py [hunk 3]
```diff
@@ -422,35 +429,79 @@ def test_find_velodrome_v2_lp_token_price(inquirer, optimism_manager):
...
+    with GlobalDBHandler().conn.write_ctx() as write_cursor:  # querying curve lp token price normally triggers curve cache query. Set all query ts to now, so it does not happen.  # noqa: E501
+        write_cursor.execute('UPDATE general_cache SET last_queried_ts=? WHERE key LIKE ?', (ts_now(), 'CURVE_LP_TOKENS%'))  # noqa: E501
+
+    inquirer.inject_evm_managers([
+        (ethereum_manager.node_inquirer.chain_id, ethereum_manager),
+        (optimism_manager.node_inquirer.chain_id, optimism_manager),
+        (polygon_pos_manager.node_inquirer.chain_id, polygon_pos_manager),
+        (arbitrum_one_manager.node_inquirer.chain_id, arbitrum_one_manager),
+        (base_manager.node_inquirer.chain_id, base_manager),
+        (gnosis_manager.node_inquirer.chain_id, gnosis_manager),
```
Comment (LefterisJP) [inline rotkehlchen/tests/unit/test_inquirer.py:497]: same logic missing here

**Review by LefterisJP — COMMENTED** [review]
There is some test failures. Also left some comments.

File: rotkehlchen/inquirer.py [hunk 4]
```diff
@@ -344,10 +351,25 @@ def inject_evm_managers(evm_managers: Sequence[tuple[ChainID, 'EvmManager']]) ->
...
+    @overload
+    @staticmethod
+    def get_evm_manager(
+            chain_id: ChainID,
+            for_curve: bool,
+    ) -> 'ArbitrumOneManager | BaseManager | EthereumManager | GnosisManager | OptimismManager | PolygonPOSManager':  # noqa: E501
+        ...
+
+    @staticmethod
+    def get_evm_manager(chain_id: ChainID, for_curve: bool = False) -> 'EvmManager':
```
Comment (LefterisJP) [inline rotkehlchen/inquirer.py:368]: What are you doing here lol. Come on. Adding an argument for type cheking? This is not how typing works. You would need to add different argument values for chain id and proper overloading. Please see how it's done in the codebase

File: rotkehlchen/inquirer.py [hunk 5]
```diff
@@ -344,10 +351,25 @@ def inject_evm_managers(evm_managers: Sequence[tuple[ChainID, 'EvmManager']]) ->
...
         evm_manager = Inquirer._evm_managers.get(chain_id)
         assert evm_manager is not None, f'evm manager for chain id {chain_id} should have been injected'  # noqa: E501
+        if for_curve:
```
Comment (LefterisJP) [inline rotkehlchen/inquirer.py:371]: Absolutely not. This is not how typing works. Remove

File: rotkehlchen/tests/unit/test_evm_misc.py [hunk 6]
```diff
@@ -55,19 +54,3 @@ def test_weth_is_supported():
         set(CHAIN_ID_TO_WETH_MAPPING.keys()) ==
         set(typing.get_args(SUPPORTED_CHAIN_IDS)) - CHAINS_WITHOUT_NATIVE_ETH
     )
-
-
-def test_evminquirer_has_cachable_mixin(blockchain):
```
Comment (LefterisJP) [inline rotkehlchen/tests/unit/test_evm_misc.py:60]: Why is this removed? Has the problem been solved? You moved the logic to the manager and not the inquirer? So this is not needed?

Comment (OjusWiZard) [reply rotkehlchen/tests/unit/test_evm_misc.py:60]: ah no, I thought it's testing something else

Comment (OjusWiZard) [reply rotkehlchen/tests/unit/test_evm_misc.py:60]: so this was testing for `ensure_cache_data_is_updated`. But now thats the part of base EvmInquirer class. So, I'm changing it to check the `CurveManagerMixin` now

File: rotkehlchen/chain/evm/node_inquirer.py [hunk 7]
```diff
@@ -173,7 +173,7 @@ def _query_web3_get_logs(
     return events
 
 
-class EvmNodeInquirer(ABC):
+class EvmNodeInquirer(ABC, LockableQueryMixIn):
```
[the remark this answers is not in the document]
Comment (OjusWiZard) [reply rotkehlchen/chain/evm/node_inquirer.py:176]: `UpdatableCacheDataMixin` is removed now right? and its function `ensure_cache_data_is_updated` is now moved to this class.
as per https://github.com/rotki/rotki/pull/8114#discussion_r1643514586

[the remark this answers is not in the document]
Comment (OjusWiZard) [reply rotkehlchen/chain/evm/node_inquirer.py:176]: Using `LockableQuery` because `ensure_cache_data_is_updated` use `@protect_with_lock` that needs the attributes defined in this mixin

File: rotkehlchen/inquirer.py [hunk 8]
```diff
@@ -819,8 +834,13 @@ def find_curve_pool_price(
         May raise:
         - RemoteError
         """
-        ethereum = self.get_evm_manager(chain_id=lp_token.chain_id)
-        ethereum.assure_curve_cache_is_queried_and_decoder_updated()  # type:ignore  # ethereum is an EthereumManager here
+        assert lp_token.chain_id in CURVE_CHAIN_IDS, f'{lp_token} is not on a curve support chain'
```
Comment (LefterisJP) [inline rotkehlchen/inquirer.py:837]:
```suggestion
        assert lp_token.chain_id in CURVE_CHAIN_IDS, f'{lp_token} is not on a curve supported chain'
```

File: rotkehlchen/inquirer.py [hunk 9]
```diff
@@ -819,8 +834,13 @@ def find_curve_pool_price(
...
+        evm_manager = self.get_evm_manager(chain_id=lp_token.chain_id)
+        assert isinstance(evm_manager, CurveManagerMixin), f'{lp_token} is not on a curve support chain'  # noqa: E501
```
Comment (LefterisJP) [inline rotkehlchen/inquirer.py:839]: Why do you need this? Doesn't it understand chain id is of type Curve chain id? Force it somehow above and please avoid this assertion

File: rotkehlchen/tests/unit/test_evm_misc.py [hunk 10]
```diff
@@ -57,17 +61,17 @@ def test_weth_is_supported():
     )
 
 
-def test_evminquirer_has_cachable_mixin(blockchain):
+def test_evmmanager_has_cachable_mixin(blockchain: 'ChainsAggregator') -> None:
```
Comment (LefterisJP) [inline rotkehlchen/tests/unit/test_evm_misc.py:64]: if typing works fine then this test is not needed

File: rotkehlchen/inquirer.py [hunk 11]
```diff
@@ -818,14 +832,20 @@ def find_curve_pool_price(
 
         May raise:
         - RemoteError
+        - AssertionError if used for a token whose chain_id is not in CURVE_CHAIN_IDS
```
Comment (LefterisJP) [inline rotkehlchen/inquirer.py:835]: AssertionError is never raised in production. `assert` statements are going off
```suggestion
```

File: rotkehlchen/inquirer.py [hunk 12]
```diff
@@ -818,14 +832,20 @@ def find_curve_pool_price(
...
         """
-        ethereum = self.get_evm_manager(chain_id=lp_token.chain_id)
-        ethereum.assure_curve_cache_is_queried_and_decoder_updated()  # type:ignore  # ethereum is an EthereumManager here
+        assert lp_token.chain_id in CURVE_CHAIN_IDS, f'{lp_token} is not on a curve support chain'
```
Comment (LefterisJP) [inline rotkehlchen/inquirer.py:837]: you did not fix the typo here
```suggestion
        assert lp_token.chain_id in CURVE_CHAIN_IDS, f'{lp_token} is not on a curve supported chain'
```

Comment (OjusWiZard) [reply rotkehlchen/inquirer.py:837]: oh I did this and then had to reset everything because many things were messed while I tried to fix the typing, and then forgot to do this again

**Review by LefterisJP — APPROVED** [review]
LGTM
